Mathlib/Geometry/Manifold/Sheaf/Smooth.lean

```diff
@@ -0,0 +1,403 @@
+  (SmoothMap.commRing : CommRing C^∞⟮IM, (unop U : Opens M); I, R⟯)
+
+/-- The presheaf of smooth functions from `M` to `R`, for `R` a smooth commutative ring, as a
+presheaf of commutative rings. -/
+def smoothPresheafCommRing : TopCat.Presheaf CommRingCat.{u} (TopCat.of M) :=
+  { obj := fun U ↦ CommRingCat.of ((smoothSheaf IM I M R).val.obj U)
+    map := fun h ↦ CommRingCat.ofHom <|
+      SmoothMap.restrictRingHom IM I R <| CategoryTheory.leOfHom h.unop
+    map_id := fun _ ↦ rfl
+    map_comp := fun _ _ ↦ rfl }
+
+/-- The sheaf of smooth functions from `M` to `R`, for `R` a smooth commutative ring, as a sheaf of
+commutative rings. -/
+def smoothSheafCommRing : TopCat.Sheaf CommRingCat.{u} (TopCat.of M) :=
+  { val := smoothPresheafCommRing IM I M R
+    cond := by
+      rw [CategoryTheory.Presheaf.isSheaf_iff_isSheaf_forget _ _
+        (CategoryTheory.forget CommRingCat)]
+      exact CategoryTheory.Sheaf.cond (smoothSheaf IM I M R) }
+
+-- sanity check: applying the `CommRingCat`-to-`TypeCat` forgetful functor to the sheaf-of-rings of
+-- smooth functions gives the sheaf-of-types of smooth functions.
+example : (CategoryTheory.sheafCompose _ (CategoryTheory.forget CommRingCat)).obj
+    (smoothSheafCommRing IM I M R) = (smoothSheaf IM I M R) := rfl
+
+instance smoothSheafCommRing.coeFun (U : (Opens (TopCat.of M))ᵒᵖ) :
+    CoeFun ((smoothSheafCommRing IM I M R).val.obj U) (fun _ ↦ ↑(unop U) → R) :=
+  (contDiffWithinAt_localInvariantProp IM I ⊤).sheafHasCoeToFun _ _ _
+
+open CategoryTheory Limits
+
+/-- Identify the stalk at a point of the sheaf-of-commutative-rings of functions from `M` to `R`
+(for `R` a smooth ring) with the stalk at that point of the corresponding sheaf of types. -/
+def smoothSheafCommRing.forgetStalk (x : TopCat.of M) :
+    (forget _).obj ((smoothSheafCommRing IM I M R).presheaf.stalk x) ≅
+    (smoothSheaf IM I M R).presheaf.stalk x :=
+  preservesColimitIso _ _
+
+@[simp, reassoc, elementwise] lemma smoothSheafCommRing.ι_forgetStalk_hom (x : TopCat.of M) (U) :
+    CategoryStruct.comp
+      (Z := (smoothSheaf IM I M R).presheaf.stalk x)
+      (FunLike.coe
+        (α := ((forget CommRingCat).obj ((smoothSheafCommRing IM I M R).presheaf.obj
+          (op ((OpenNhds.inclusion x).obj U.unop)))))
+        (colimit.ι ((OpenNhds.inclusion x).op ⋙ (smoothSheafCommRing IM I M R).presheaf) U))
+      (forgetStalk IM I M R x).hom =
+    colimit.ι ((OpenNhds.inclusion x).op ⋙ (smoothSheaf IM I M R).presheaf) U :=
+  ι_preservesColimitsIso_hom _ _ _
+
+@[simp, reassoc, elementwise] lemma smoothSheafCommRing.ι_forgetStalk_inv (x : TopCat.of M) (U) :
+    colimit.ι ((OpenNhds.inclusion x).op ⋙ (smoothSheaf IM I M R).val) U ≫
+    (smoothSheafCommRing.forgetStalk IM I M R x).inv =
+    (forget CommRingCat).map
+      (colimit.ι ((OpenNhds.inclusion x).op ⋙ (smoothSheafCommRing IM I M R).val) U) := by
+  rw [Iso.comp_inv_eq, ← smoothSheafCommRing.ι_forgetStalk_hom]
+  rfl
```

Comment:
I think it's better not to lean so hard on `rfl` in situations like this:
```suggestion
  rw [Iso.comp_inv_eq, ← smoothSheafCommRing.ι_forgetStalk_hom, CommRingCat.forget_map]
  simp_rw [Functor.comp_obj, Functor.op_obj]
```
or even
```lean
  rw [← smoothSheafCommRing.ι_forgetStalk_hom]
  simp
```